openlabcmd/openlabcmd/zk.py

```diff
@@ -206,37 +217,116 @@ def delete_node(self, node_name):
         self.client.delete(path, recursive=True)
 
     @_client_check_wrapper
-    def list_services(self):
+    def list_services(self, node_name_filter=None, node_role_filter=None):
+        """
+        List the services in the HA deployment.
+        :param node_name_filter: The node filter.
+        :type node_name_filter: list or string.
+        :param node_role_filter: The node filter.
+        :type node_role_filter: list or string.
+        :return: the services list.
+        """
+        if node_name_filter:
+            if isinstance(node_name_filter, str):
+                node_name_filter = [node_name_filter]
+            if not isinstance(node_name_filter, list):
+                raise exceptions.ValidationError("node_name_filter should be "
+                                                 "a list or string.")
+        if node_role_filter:
+            if isinstance(node_role_filter, str):
+                node_role_filter = [node_role_filter]
+            if not isinstance(node_role_filter, list):
+                raise exceptions.ValidationError("node_role_filter should be "
+                                                 "a list or string.")
+
         result = []
         for exist_node in self.list_nodes():
+            if node_name_filter and exist_node['name'] not in node_name_filter:
+                continue
+            if node_role_filter and exist_node['role'] not in node_role_filter:
+                continue
             path = '/ha/%s/%s_service' % (exist_node['name'],
                                           exist_node['role'])
             for service_name in self.client.get_children(path):
                 service_path = path + '/' + service_name
                 service_info = self.client.get(service_path)
-                service_info = self._bytesToDict(service_info[0])
+                service_info = self._bytesToDict(service_info,
+                                                 include_updated_at=True)
```

Comment:
Just a question: when the update_at is changed for service, there is not set or update operate for service, right? if it's changed, why created_at is not needed here?

Reply:
sorry, if it's can be changed.....

Reply:
sorry, misunderstand, ignore this.